File: SG_UDF/REGEX.fs
namespace FSUDF_REGEX
open System
open System.Data
open System.Data.SqlClient
open System.Data.SqlTypes
open Microsoft.SqlServer.Server
open System.Collections
open System.Collections.Generic
open System.Text.RegularExpressions

module regexUtil =
    type regexAct =
    | Match = 0
    | Replace = 1
    | Matches = 2

    let regexFun = fun target expr group regexActMode replStr replCnt replStart (o : option<ref<_>>) ->
        let (|M|R|Ms|) regexActMode =
            match regexActMode % 3 with
             | 0 -> M
             | 1 -> R
             | 2 -> Ms
        let ra =
            match regexActMode with
            | M ->
                enum<regexAct>(0)
            | R ->
                enum<regexAct>(1)
            | Ms ->
                enum<regexAct>(2)
        let g = group :> Object
        let gn = ref 0
        let regex = new Regex(expr, RegexOptions.Multiline + RegexOptions.IgnoreCase)
        match ra with
        | regexAct.Match ->
            let varMatch = regex.Match(target)
            match varMatch.Success with
            | false -> null
            | true ->
                match Array.Exists(regex.GetGroupNames(), (fun gpnm -> (gpnm = (string)g))) with
                | true ->
                    varMatch.Groups.[(string)g].Captures.[0].Value
                | _ when Int32.TryParse((string)g, gn) ->
                    varMatch.Groups.Item(gn.Value).Captures.[0].Value
            | _ -> varMatch.Groups.[0].Captures.[0].Value
        | regexAct.Replace ->
            regex.Replace(target, (string)replStr, replCnt, replStart)
        | regexAct.Matches ->
            let ms = regex.Matches(target)
            o.Value := ms
            "0"
type public SPParameters =
    class
        [<DefaultValue>] val mutable ParamID : SqlInt32
        [<DefaultValue>] val mutable ParamName : SqlString
        [<DefaultValue>] val mutable ParamType : SqlString
        [<DefaultValue>] val mutable ParamDefaultStr : SqlString
        [<DefaultValue>] val mutable ParamDefaultValue : SqlString
        new () ={}
    end

type public UDF_REGEX =
    static member FillSPParameters(o : obj
        , param_id : SqlInt32 byref
        , param_name : SqlString byref
        , param_type : SqlString byref
        , param_default_str : SqlString byref
        , param_default_value : SqlString byref) =
        let o' = o :?> SPParameters
        param_id <- o'.ParamID
        param_name <- o'.ParamName
        param_type <- o'.ParamType
        param_default_str <- o'.ParamDefaultStr
        param_default_value <- o'.ParamDefaultValue
    [<SqlFunction(
        DataAccess = DataAccessKind.Read,
        FillRowMethodName = "FillSPParameters", 
        TableDefinition = "
            [ParamID] [int],
            [ParamName] [nvarchar](50),
            [ParamType] [nvarchar](50),
            [ParamDefaultValueSQLStr] [nvarchar](max),
            [ParamDefaultValue] [nvarchar](max)")>]
        static member GetSPParamDefaultValue(spDefinition : string) =
            let results = new ArrayList()
            //let obj = new Object()
            let o = Some (ref (Regex.Matches("", "")))
            let expr =
                "\s*@(?<nm>\w+)\s*" +
                "(?<type>[^\s\(]+\s*\(\s*((\w+\s*\))|(\d+\s*\))|(\d+\s*,\s*\d+\s*\))))\s*=\s*" +
                "(?<defvalueStr>((?<!(?-i)N*')(?<defvalue>\w+)\s*)|((?-i)N*'" +
                "(?<defvalue>(((?<![^']')''(?!'[^']))|([^']))*)" +
                "'))\s*(,|\))"
            regexUtil.regexFun spDefinition expr null 2 null 0 0 o |> ignore
            let mc = !o.Value // :> System.Text.RegularExpressions.MatchCollection
            let mutable idx = 0
            for m in mc do
                let spparam' = new SPParameters()
                spparam'.ParamID <- new SqlInt32(idx)
                spparam'.ParamName <- new SqlString(m.Groups.["nm"].Value)
                spparam'.ParamType <- new SqlString(m.Groups.["type"].Value)
                spparam'.ParamDefaultStr <- new SqlString(m.Groups.["defvalueStr"].Value)
                spparam'.ParamDefaultValue <- new SqlString(m.Groups.["defvalue"].Value)
                idx <- idx + 1
                results.Add(spparam') |> ignore
            results
 
    [<SqlFunction(IsDeterministic = true, IsPrecise = true)>]
        static member regexMatch(target : string, expr : string) =
            let regex = new Regex(expr, RegexOptions.Multiline + RegexOptions.IgnoreCase)
            regex.IsMatch(target)

    [<SqlFunction(IsDeterministic = true, IsPrecise = true)>]
        static member regexVarMatchNm(target : string, expr : string, group : string) =
            (string)(regexUtil.regexFun target expr group 0 null 0 0 None)

    [<SqlFunction(IsDeterministic = true, IsPrecise = true)>]
        static member regexVarMatchId(target : string, expr : string, group : int) =
            (string)(regexUtil.regexFun target expr group 0 null 0 0 None)
 
    [<SqlFunction(IsDeterministic = true, IsPrecise = true)>]
        static member regexReplace(target : string, expr : string, replacement : string) =
            (string)(regexUtil.regexFun target expr null 1 replacement -1 0 None)

File: CLR_UDF_CS/REGEX.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;


namespace CSUDF_REGEX {
    public static class regexUtil {

        public enum regexAct {
            Match = 0,
            Replace = 1,
            Matches = 2,
        };

        public static string regexFun(string target, string expr, object g, int rai, string replStr, int replCnt, int replStart, ref MatchCollection o, int capID)
        {
            regexAct ra = (regexAct)rai;
            return regexFun(target, expr, g, ra, replStr, replCnt, replStart, ref o, capID);
        }

        public static string regexFun(string target, string expr, object g, int rai, string replStr, int replCnt, int replStart, int capID)
        {
            regexAct ra = (regexAct)rai;
            return regexFun(target, expr, g, ra, replStr, replCnt, replStart, capID);
        }

        public static string regexFun(string target, string expr, object g, regexAct ra, string replStr, int replCnt, int replStart, ref MatchCollection o, int capID) {
            int gn = g is int ? Int32.Parse(g.ToString()) : 0;
            if (capID < 0) { capID = 0; }
            Regex regex = new Regex(expr, RegexOptions.Multiline | RegexOptions.IgnoreCase);
            try {
                switch (ra) {
                    case regexAct.Match:
                        var varMatch = regex.Match(target);
                        if (!varMatch.Success) {
                            return null;
                        } else if (g is String && Array.Exists(regex.GetGroupNames(), gpnm => (gpnm == g.ToString()))) {
                            return varMatch.Groups[g.ToString()].Captures[capID].Value;
                        } else if (g is int || Int32.TryParse(g.ToString(), out gn)) {
                            return varMatch.Groups[gn].Captures[capID].Value;
                        } else {
                            return varMatch.Groups[0].Captures[capID].Value;
                        }
                    case regexAct.Replace:
                        return regex.Replace(target, (string)replStr, replCnt, replStart);
                    case regexAct.Matches:
                        var ms = regex.Matches(target);
                        o = ms;
                        return "0";
                    default:
                        return "err:-1";
                }
            } catch{
                return "err:-2";
            }
        }

        public static string regexFun(string target, string expr, object g, regexAct ra, string replStr, int replCnt, int replStart, int capID) {
            var o = Regex.Matches("", "");
            return regexFun(target, expr, g, ra, replStr, replCnt, replStart, ref o, capID);
        }
    }

    public class SPParameters {
        public SqlInt32 ParamID;
        public SqlString ParamName;
        public SqlString ParamType;
        public SqlString ParamDefaultStr;
        public SqlString ParamDefaultValue;
    }

    public static class UDF_REGEX {
        public static void FillSPParameters(
            object o
            , ref SqlInt32 param_id
            , ref SqlString param_name
            , ref SqlString param_type
            , ref SqlString param_default_str
            , ref SqlString param_default_value) {
            var oo = (SPParameters)o;
            param_id = oo.ParamID;
            param_name = oo.ParamName;
            param_type = oo.ParamType;
            param_default_str = oo.ParamDefaultStr;
            param_default_value = oo.ParamDefaultValue;
        }
        [SqlFunction(
            DataAccess = DataAccessKind.Read,
            FillRowMethodName = "FillSPParameters",
            TableDefinition = @"
                [ParamID][int],
                [ParamName][nvarchar](50),
                [ParamType][nvarchar](50),
                [ParamDefaultValueSQLStr][nvarchar](max),
                [ParamDefaultValue][nvarchar](max)")]
        public static ArrayList GetSPParamDefaultValue(string spDefinition) {
            var results = new ArrayList();
            MatchCollection o = null;
            string expr =
                "\\s*@(?<nm>\\w+)\\s*" +
                "(?<type>[^\\s\\(]+\\s*\\(\\s*((\\w+\\s*\\))|(\\d+\\s*\\))|(\\d+\\s*,\\s*\\d+\\s*\\))))\\s*=\\s*" +
                "(?<defvalueStr>((?<!(?-i)N*')(?<defvalue>\\w+)\\s*)|((?-i)N*'" +
                "(?<defvalue>(((?<![^']')''(?!'[^']))|([^']))*)" +
                "'))\\s*(,|\\))";
            regexUtil.regexFun(spDefinition, expr, null, (regexUtil.regexAct)2, null, 0, 0, ref o, -1);
            int idx = 0;
            foreach (Match m in o) {
                var spparam = new SPParameters();
                spparam.ParamID = new SqlInt32(idx);
                spparam.ParamName = new SqlString(m.Groups["nm"].Value);
                spparam.ParamType = new SqlString(m.Groups["type"].Value);
                spparam.ParamDefaultStr = new SqlString(m.Groups["defvalueStr"].Value);
                spparam.ParamDefaultValue = new SqlString(m.Groups["defvalue"].Value);
                idx++;
                results.Add(spparam);
            }
            return results;
        }
        [SqlFunction(IsDeterministic = true, IsPrecise = true)]
        public static bool regexMatch(string target, string expr) {
            var regex = new Regex(expr, RegexOptions.Multiline & RegexOptions.IgnoreCase);
            return regex.IsMatch(target);
        }
        [SqlFunction(IsDeterministic = true, IsPrecise = true)]
        public static string regexVarMatchNm(string target, string expr, string group) {
            return (string)regexUtil.regexFun(target, expr, group, 0, null, 0, 0, 0);
        }
        [SqlFunction(IsDeterministic = true, IsPrecise = true)]
        public static string regexVarMatchId(string target, string expr, int group) {
            return (string)regexUtil.regexFun(target, expr, group, 0, null, 0, 0, 0);
        }   
        [SqlFunction(IsDeterministic = true, IsPrecise = true)]
        public static string regexReplace(string target, string expr, string replacement) {
            return (string)regexUtil.regexFun(target, expr, null, 1, replacement, -1, 0, -1);
        }

        [SqlFunction(IsDeterministic = true, IsPrecise = true)]
        public static string regexVarMatchNm2(string target, string expr, string group, int capID)
        {
            return (string)regexUtil.regexFun(target, expr, group, 0, null, 0, 0, capID);
        }
        [SqlFunction(IsDeterministic = true, IsPrecise = true)]
        public static string regexVarMatchId2(string target, string expr, int group, int capID)
        {
            return (string)regexUtil.regexFun(target, expr, group, 0, null, 0, 0, capID);
        }
    }
}